### PR TITLE
Handle Recursive inheritance and model circle reference

### PR DESCRIPTION
**Pull request by chunyu3**
# Description
Fix https://github.com/Azure/autorest.csharp/issues/4016
Fix https://github.com/Azure/autorest.csharp/issues/3974
Fix https://github.com/Azure/autorest.csharp/issues/4029

- remove duplicated model schema in tspCodeModel.json
- Add visistedModel to skip visisted model in FindMatchesRecursively() to avoid infinite loop when there is a circle reference in the model.

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

**Comment by pshao25** [comment]
Do we have tests?

File: src/AutoRest.CSharp/Common/Input/InputTypes/Serialization/TypeSpecReferenceHandler.cs [hunk 1]
```diff
@@ -24,9 +24,9 @@ private class TypeSpecReferenceResolver : ReferenceResolver
 
             public override void AddReference(string referenceId, object value)
             {
-                if (!_referenceIdToObjectMap.TryAdd(referenceId, value))
+                if (!_referenceIdToObjectMap.ContainsKey(referenceId))
                 {
-                    throw new JsonException();
+                    _referenceIdToObjectMap.Add(referenceId, value);
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Input/InputTypes/Serialization/TypeSpecReferenceHandler.cs:29]: this scenario "adding a reference when the same Id existed" should never happen, therefore I believe we should revert this change.
We should make sure that definition of a model should only happen once.

Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Input/InputTypes/Serialization/TypeSpecReferenceHandler.cs:29]: No. we cannot make the model.derievedModels only once otherwise we will meet circle loop issue of the json deserialization issue.

**Comment by chunyu3** [comment]
> Do we have tests?

hello @pshao25 we will have test in latest cadl-ranch. After upgraded cadl-ranch, I will onboard the test.

File: src/AutoRest.CSharp/Common/Input/InputTypes/Serialization/TypeSpecReferenceHandler.cs (hunk 1 above)
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Input/InputTypes/Serialization/TypeSpecReferenceHandler.cs:29]: Let us add a TODO here to emphasize this is a workaround.

The solution of this issue should be in the emitter to ensure it will not output the derived type twice and we should NEVER get a duplicated key here

**Review by ArcturusZhang — APPROVED** [review]
Approved if this is only a workaround and there is follow up to fix the real issue.

**Comment by ArcturusZhang** [comment]
> > Do we have tests?
> 
> hello @pshao25 we will have test in latest cadl-ranch. After upgraded cadl-ranch, I will onboard the test.

I suggest we onboard the test case now, and when the cadl ranch case is onboarded we could have another PR to remove it.

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs [hunk 2]
```diff
@@ -6,9 +6,14 @@
...
+    internal record InputModelType(string Name, string? Namespace, string? Accessibility, string? Deprecated, string? Description, InputModelTypeUsage Usage, IReadOnlyList<InputModelProperty> Properties, IReadOnlyList<InputModelType> DerivedModels, string? DiscriminatorValue, string? DiscriminatorPropertyName, bool IsNullable)
         : InputType(Name, IsNullable)
     {
+        public InputModelType(string name, string? ns, string? accessibility, string? deprecated, string? description, InputModelTypeUsage usage, IReadOnlyList<InputModelProperty> properties, InputModelType? baseModel, IReadOnlyList<InputModelType> derivedModels, string? discriminatorValue, string? discriminatorPropertyName, bool isNullable)
+            : this(name, ns, accessibility, deprecated, description, usage, properties, derivedModels, discriminatorValue, discriminatorPropertyName, isNullable)
+        {
+            BaseModel = baseModel;
+        }
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:16]: these changes are unacceptable.
According to our discussion yesterday, these classes (this one and the corresponding converter/deserializer) do not need to be changed, it is already possible to handle the case we are currently having.
The only issue is in the emitter.
If we would like to introduce a temporary workaround, I prefer the previous way which is simpler and straightforward to revert.

File: test/TestProjects/FirstTest-TypeSpec/src/Generated/tspCodeModel.json [hunk 3]
```diff
@@ -748,19 +748,146 @@
...
+      "IsNullable": false
+     },
+     "IsRequired": false,
+     "IsReadOnly": false
+    }
+   ]
+  },
+  {
+   "$ref": "105"
+  },
```
Comment (ArcturusZhang) [inline test/TestProjects/FirstTest-TypeSpec/src/Generated/tspCodeModel.json:801]: this is exactly the result we would like to see in the emitter, and in our yesterday's discussion, it does not require a change in the converter to work properly.

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs (hunk 2 above)
Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:16]: And I have a try on my machine, by checking out to your branch, revert all the changes in CodeModelConverter.cs, InputModelType.cs, and TypeSpecInputModelTypeConverter.cs, then regenerate FirstTest-TypeSpec, it works fine and generated code does not have a change.
Is there anything I am missing that we really require these changes?

Comment (ArcturusZhang) [reply src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:16]: I put the changes in this PR: https://github.com/Azure/autorest.csharp/pull/4038

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs [hunk 4]
```diff
@@ -3,12 +3,19 @@
...
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:18]: We do not have to change the first line to add a private setter to BaseModel property, we could do something like this:
image
this might reduce the amount of changes in this PR

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs [hunk 5]
```diff
@@ -19,6 +26,13 @@ internal record InputModelType(string Name, string? Namespace, string? Accessibi
         public bool IsPropertyBag { get; init; } = false;
 
         public bool IsAnonymousModel { get; init; } = false;
+        public InputModelType? BaseModel { get; private set; }
+        internal void SetBaseModel(InputModelType? baseModel, [CallerFilePath] string filepath = "", [CallerMemberName] string caller = "")
+        {
+            Debug.Assert(filepath.EndsWith($"{nameof(TypeSpecInputModelTypeConverter)}.cs"));
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:32]: Could we add a message here as well?

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs [hunk 6]
```diff
@@ -19,6 +26,13 @@ internal record InputModelType(string Name, string? Namespace, string? Accessibi
         public bool IsPropertyBag { get; init; } = false;
 
         public bool IsAnonymousModel { get; init; } = false;
+        public InputModelType? BaseModel { get; private set; }
+        internal void SetBaseModel(InputModelType? baseModel, [CallerFilePath] string filepath = "", [CallerMemberName] string caller = "")
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:30]: Could we add some xml document here to explain the reasoning of doing this?

File: src/AutoRest.CSharp/Common/Input/InputTypes/Serialization/TypeSpecInputModelTypeConverter.cs [hunk 7]
```diff
@@ -51,17 +51,27 @@ public static InputModelType CreateModelType(ref Utf8JsonReader reader, string?
...
-                    || reader.TryReadWithConverter(nameof(InputModelType.BaseModel), options, ref baseModel);
+                    || reader.TryReadString(nameof(InputModelType.DiscriminatorValue), ref discriminatorValue);
 
                 if (isKnownProperty)
                 {
                     continue;
                 }
-
-                if (reader.GetString() == nameof(InputModelType.Properties))
+                if (reader.GetString() == nameof(InputModelType.BaseModel))
```
Comment (ArcturusZhang) [inline src/AutoRest.CSharp/Common/Input/InputTypes/Serialization/TypeSpecInputModelTypeConverter.cs:60]: we should add some comments here to emphasize the last two properties in the JSON must be `BaseModel` and `Properties` and must in this order

**Review by ArcturusZhang — APPROVED** [review]

File: src/AutoRest.CSharp/Common/Input/InputTypes/Serialization/TypeSpecInputModelTypeConverter.cs (hunk 7 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Input/InputTypes/Serialization/TypeSpecInputModelTypeConverter.cs:64]: added.

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs (hunk 6 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs:35]: added.

File: src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs (hunk 5 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Input/InputTypes/InputModelType.cs]: added.

**Review by lirenhe — APPROVED** [review]
